sequencing-server/src/utils/hasura.ts

```diff
@@ -0,0 +1,207 @@
+    gql`
+      query checkUserPermissions($role: user_roles_enum!, $key: String!) {
+        user_role_permission_by_pk(role: $role) {
+          action_permissions(path: $key)
+        }
+      }
+    `,
+    {
+      role,
+      key: ENDPOINTS_TO_ACTION_KEY[url],
+    },
+  );
+
+  const permission = permissionCheckQuery.user_role_permission_by_pk.action_permissions as HasuraPermissions | null;
+
+  if (permission === null) {
+    return false;
+  }
+
+  const missionModelId = body.input.missionModelId as number;
+  const planId = body.input.planId as number;
+
+  switch (permission) {
+    case HasuraPermissions.NO_CHECK:
+      return true;
+    case HasuraPermissions.MISSION_MODEL_OWNER:
+      return missionModelId !== null && (await isMissionModelOwner(graphqlClient, missionModelId, user));
+    case HasuraPermissions.OWNER:
+      return planId !== null && (await isPlanOwner(graphqlClient, planId, user));
+    case HasuraPermissions.PLAN_OWNER:
+      return planId !== null && (await isPlanOwner(graphqlClient, planId, user));
+    case HasuraPermissions.PLAN_COLLABORATOR:
+      return planId !== null && (await isPlanCollaborator(graphqlClient, planId, user));
+    case HasuraPermissions.PLAN_OWNER_COLLABORATOR:
+      return (
+        (planId !== null && (await isPlanOwner(graphqlClient, planId, user))) ||
+        (await isPlanCollaborator(graphqlClient, planId, user))
+      );
+  }
+}
+
+async function isMissionModelOwner(
+  graphqlClient: GraphQLClient,
+  missionModelId: number,
+  user: string,
+): Promise<boolean> {
+  const missionModelOwner = await graphqlClient.request<{
+    mission_model: { owner: string }[];
+  }>(
+    gql`
+      query missionModelOwner($missionModelId: Int!) {
+        mission_model(where: { id: { _eq: $missionModelId } }) {
+          owner
+        }
+      }
+    `,
+    { missionModelId },
+  );
+
+  return user === missionModelOwner.mission_model[0]?.owner;
+}
+
+async function isPlanOwner(graphqlClient: GraphQLClient, planId: number, user: string): Promise<boolean> {
+  const planOwner = await graphqlClient.request<{
+    plan: { owner: string }[];
+  }>(
+    gql`
+      query planOwner($planId: Int!) {
+        plan(where: { id: { _eq: $planId } }) {
+          owner
+        }
+      }
+    `,
+    { planId },
+  );
+
+  return user === planOwner.plan[0]?.owner;
+}
+
+async function isPlanCollaborator(graphqlClient: GraphQLClient, planId: number, user: string): Promise<boolean> {
+  const planCollaboratorRes = await graphqlClient.request<{
+    plan_collaborators: { collaborator: string }[];
+  }>(
+    gql`
+      query planCollaborator($planId: Int!) {
+        plan_collaborators(where: { id: { _eq: $planId } }) {
+          collaborator
+        }
+      }
+    `,
+    { planId },
+  );
+
+  for (const planCollaborator of planCollaboratorRes.plan_collaborators) {
+    if (user === planCollaborator.collaborator) {
+      return true;
+    }
+  }
```

Comment:
This query will simplify the function, as it will either return `null` or `$username`
```gql
query planCollaborator($planId: Int!, $username: String!) {
  plan_collaborators_by_pk(plan_id: $planId, collaborator: $username) {
    collaborator
  }
}
```